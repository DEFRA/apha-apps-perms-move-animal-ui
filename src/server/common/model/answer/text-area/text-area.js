import Joi from 'joi'
import { AnswerModel } from '../answer-model.js'
import { validateAnswerAgainstSchema } from '../validation.js'
import { NotImplementedError } from '../../../helpers/not-implemented-error.js'

/** @import {AnswerViewModelOptions} from '../answer-model.js' */

/**
 * @param {TextAreaConfig} config
 * @returns {Joi.Schema}
 */
const textAreaSchema = ({ payloadKey, validation }) => {
  let stringValidation = Joi.string()
    .required()
    .trim()
    .max(validation.maxLength.value)

  const messages = {
    'string.max': validation.maxLength.message
  }

  if (validation.empty) {
    messages['any.required'] = validation.empty.message
    messages['string.empty'] = validation.empty.message
  } else {
    stringValidation = stringValidation.allow('')
  }

  if (validation.pattern) {
    stringValidation = stringValidation.pattern(validation.pattern.regex)
    messages['string.pattern.base'] = validation.pattern.message
  }

  return Joi.object({
    [payloadKey]: stringValidation.messages(messages)
  })
}

/**
 * export @typedef {{
 *  payloadKey: string,
 *  spellcheck?: false,
 *  rows?: number,
 *  autocomplete?: string,
 *  hint?: string,
 *  validation: {
 *    maxLength: { value: number, message: string },
 *    empty?: { message: string },
 *    pattern?: { regex: RegExp, message: string }
 *  }
 * }} TextAreaConfig
 */

/**
 * @typedef {string} TextAreaData
 */

/**
 * @template Payload
 * @augments {AnswerModel<Payload>}
 */
export class TextAreaAnswer extends AnswerModel {
  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {TextAreaConfig} */
  get config() {
    return /** @type {any} */ (this.constructor).config
  }

  // eslint-disable-next-line jsdoc/require-returns-check
  /** @returns {TextAreaConfig} */
  static get config() {
    throw new NotImplementedError()
  }

  /**
   * @returns {string | undefined}
   */
  get value() {
    return this._data?.[this.config.payloadKey]
  }

  get html() {
    return this._data?.[this.config.payloadKey]?.replace(/\n/g, '<br />') ?? ''
  }

  // eslint-disable-next-line @typescript-eslint/class-literal-property-style
  get template() {
    return 'model/answer/text-area/text-area.njk'
  }

  /**
   * @returns {TextAreaData}
   */
  toState() {
    return this.value?.trim() ?? ''
  }

  validate() {
    return validateAnswerAgainstSchema(
      textAreaSchema(this.config),
      this._data ?? {}
    )
  }

  /**
   * @param {AnswerViewModelOptions} options
   */
  viewModel({ validate, question }) {
    const { payloadKey, spellcheck, autocomplete, rows, hint } = this.config
    const viewModel = {
      label: {
        text: question,
        classes: 'govuk-label--l',
        isPageHeading: true
      },
      id: payloadKey,
      name: payloadKey,
      value: this.value
    }

    if (rows) {
      viewModel.rows = rows
    }

    if (autocomplete) {
      viewModel.autocomplete = autocomplete
    }

    if (hint) {
      viewModel.hint = { text: hint }
    }

    if (spellcheck === false) {
      viewModel.spellcheck = false
    }

    if (validate) {
      viewModel.errorMessage = this.validate().errors[payloadKey]
    }

    return viewModel
  }

  /**
   * @param {Payload} fields
   */
  _extractFields(fields) {
    return /** @type {Payload} */ ({
      [this.config.payloadKey]: fields[this.config.payloadKey]
    })
  }

  /**
   * @param {TextAreaData | undefined} state
   * @returns {TextAreaAnswer}
   */
  static fromState(state) {
    return new this(
      state !== undefined ? { [this.config.payloadKey]: state } : {}
    )
  }
}
