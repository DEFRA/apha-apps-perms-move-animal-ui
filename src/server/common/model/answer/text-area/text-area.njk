{% from "govuk/components/textarea/macro.njk" import govukTextarea %}
{% from "govuk/components/fieldset/macro.njk" import govukFieldset %}

{% call govukFieldset() %}
  {{ govukTextarea(answerViewModel) }}
{% endcall %}
