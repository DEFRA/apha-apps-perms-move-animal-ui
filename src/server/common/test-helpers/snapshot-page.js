import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
