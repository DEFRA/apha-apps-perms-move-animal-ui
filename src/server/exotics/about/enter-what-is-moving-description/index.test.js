import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { Answer, enterWhatIsMovingDescriptionPage } from './index.js'
import { TextAreaAnswer } from '~/src/server/common/model/answer/text-area/text-area.js'
import { CheckAnswersPage } from '../check-answers/index.js'

const sectionKey = 'about'
const questionKey = 'whatAreYouMovingDescrition'
const pageUrl =
  '/exotics/about-the-movement/what-is-moving/enter-what-is-moving/description'
const page = enterWhatIsMovingDescriptionPage

const payload = {
  [questionKey]: 'some text'
}

describe('Answer', () => {
  it('should be a text area input', () => {
    expect(new Answer(payload)).toBeInstanceOf(TextAreaAnswer)
  })

  it('should have the right payload key', () => {
    expect(Answer.config.payloadKey).toBe(questionKey)
  })

  it('should define the right empty input message', () => {
    expect(Answer.config.validation.empty?.message).toBe(
      'Enter information on what you are moving'
    )
  })

  it('should define the right max lengnth input message', () => {
    expect(Answer.config.validation.maxLength).toStrictEqual({
      value: 5000,
      message: 'Your answer must be no longer than 5000 characters'
    })
  })
})

describe('EnterWhatIsMovingDescriptionPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  describe('nextPage', () => {
    it('should return EnterWhatIsMovingDescriptionPage for any value', () => {
      const nextPage = page.nextPage()
      expect(nextPage).toBeInstanceOf(CheckAnswersPage)
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
