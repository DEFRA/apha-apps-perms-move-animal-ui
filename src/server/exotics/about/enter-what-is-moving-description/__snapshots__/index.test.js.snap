// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterWhatIsMovingDescriptionPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <h1 class="govuk-label-wrapper">
    <label class="govuk-label govuk-label--l" for="whatAreYouMovingDescription">
      Describe what you are moving
    </label>
  </h1>
  <textarea class="govuk-textarea" id="whatAreYouMovingDescription" name="whatAreYouMovingDescription" rows="10" spellcheck="false"></textarea>
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
